### PR TITLE
[BEAM-347] Modify range tracker to use first response as start key

### PR DESCRIPTION
**Pull request by ianzhou1**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @dhalperi

**Comment by dhalperi** [comment]
This PR contains a lot of other commits, it looks like. Can you please rebase on top of master and force-push?

This assumes you followed the `remote` naming scheme from the Beam contribution guide:

```sh
git fetch --all
git rebase apache/master
git log apache/master..    # this should only show your CL
git push -f ianzhou1 HEAD:beam
```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java [hunk 1]
```diff
@@ -767,6 +767,9 @@ public boolean start() throws IOException {
           reader.start()
               && rangeTracker.tryReturnRecordAt(true, ByteKey.of(reader.getCurrentRow().getKey()));
       if (hasRecord) {
+        if (recordsReturned == 0) {
+          source = source.withStartKey(rangeTracker.getStartPosition());
+        }
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java]: I don't think this code needs to change. The range tracker could handle all of the range adjustment. Can you see why? (Maybe there's an issue somewhere else in BigtableIO?)

**Comment by kennknowles** [comment]
I see you have BEAM-??? in your title. Is there a ticket filed about this, so I can gain context on it? Just for my own edification.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java [hunk 2]
```diff
@@ -811,17 +811,19 @@ public final Double getFractionConsumed() {
...
+            "%s: Failed to interpolate key for fraction %s.", rangeTracker.getRange(), fraction);
         return null;
       }
       logger.debug(
           "Proposing to split {} at fraction {} (key {})", rangeTracker, fraction, splitKey);
       if (!rangeTracker.trySplitAtPosition(splitKey)) {
         return null;
       }
-      BigtableSource primary = source.withEndKey(splitKey);
+      BigtableSource primary =
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java]: This is subtle, but I don't think that we want this change. If the Source's range is `[a,z)` and the first key is `d`, I think the source's range is still `[a, z)`. Leave the start key alone, but the rangeTracker will still keep the state the same.

(The change above is still good)

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java [hunk 3]
```diff
@@ -25,7 +25,6 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Verify.verifyNotNull;
-
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java]: please restore.

We may need to update the intellij config I sent you if it was removing this line

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java [hunk 4]
```diff
@@ -28,63 +28,78 @@
...
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
 
-    assertEquals(START_KEY, tracker.getStartPosition());
+    assertEquals(INITIAL_START_KEY, tracker.getStartPosition());
     assertEquals(END_KEY, tracker.getStopPosition());
   }
 
   /** Tests for {@link ByteKeyRangeTracker#getFractionConsumed()}. */
   @Test
   public void testGetFractionConsumed() {
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java]: I expected a test of `getFractionConsumed` with a key that is NOT the initial start key, to ensure that progress gets and stays updated.

**Comment by dhalperi** [comment]
Looks pretty good, Ian. Lots of comments, but all details.

Please do file a JIRA issue explaining the bug and then update the issue number in the title.

Happy to sit down with you to draft the issue desc if you're not comfortable writing from scratch, but you should go for it! Good intern task. ;)

**Comment by ianzhou1** [comment]
ptal @dhalperi

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java [hunk 5]
```diff
@@ -28,66 +28,96 @@
...
+    assertEquals(0.0, tracker.getFractionConsumed(), delta);
+
+    tracker.tryReturnRecordAt(true, INITIAL_MIDDLE_KEY);
+    assertEquals(0.5, tracker.getFractionConsumed(), delta);
+
+    tracker.tryReturnRecordAt(true, BEFORE_END_KEY);
+    assertEquals(1 - 1 / INITIAL_RANGE_SIZE, tracker.getFractionConsumed(), delta);
+
+    // Should be accurate for updated start key
+    tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java]: please make it a separate test.

`testGetFractionConsumed` as it was, and then `testGetFractionConsumedLateStartKey`, for example

**Comment by dhalperi** [comment]
One last comment.

**Comment by ianzhou1** [comment]
ptal again

**Comment by dhalperi** [comment]
LGTM. I can merge once the tests are green.

**Pull request closed without merging**